### PR TITLE
[Dialog][AlertDialog][Popover][DropdownMenu][ContextMenu] Modality

### LINKED ISSUE #566: [DropdownMenu] Can't enable scroll for `DropdownMenu` in overflow scroll container
## Bug report

`<DropdownMenu.Content disableOutsideScroll={false}>` doesn't do anything when the Dropdown is inside a scrollable container
https://codesandbox.io/s/gifted-wave-xg810

### LINKED ISSUE #585: [Dialog] Consider adding `disableOutsidePointerEvents`
### Overview

I need to remove `disableOutsidePointerEvents` feature in `Dialog` component but it seems impossible  https://github.com/radix-ui/primitives/blob/main/packages/react/dialog/src/Dialog.tsx#L241. 

### Additional context

Such option is already exists in DropdownMenu.Content component, would be useful to have it in Dialog as well.

### PR DESCRIPTION
**Pull request by benoitgrelard**
Closes #585
Closes #566

This PR adds modality support to menus, dialogs and popovers. Originally we exposed a very granular api for managing pointer events, scroll locks and focus control. The most common use case for these props was to apply differing levels of modality depending on consumer needs.

By consolidating and streamlining these into a single `modal` prop we are able to reduce the api surface while making the outcome easier and faster to achieve. This change also allows us to provide consistent handling across components and improve some additional handling under the hood.

### Example (DropdownMenu)

https://user-images.githubusercontent.com/11708259/125466947-c6f22aaa-c5be-4546-ae5e-b8acacfd477e.mp4

### CODE REVIEW

File: packages/react/menu/src/Menu.tsx
Comment (jjenzz) [inline packages/react/menu/src/Menu.tsx:300]: I need a refresher on this one. I can see this switches off trapping but doesn't enable looping. 

@benoitgrelard I thought we had said that if things were non-modal they would `loop` and not `trap` so that we didn't have to worry about moving focus to the element after the trigger?

Comment (jjenzz) [reply packages/react/menu/src/Menu.tsx:300]: Ah I can see we do `loop` in popover. I'm guessing we don't here because it's a menu and that's all handled by `RovingFocusGroup` anyway?

Comment (andy-hook) [reply packages/react/menu/src/Menu.tsx:300]: >  I'm guessing we don't here because it's a menu which tend not to loop

That was my understanding. Tab's currently [blocked on menus](https://github.com/radix-ui/primitives/blob/80dacf836f3be65941d0f3fa017991f75b1313a2/packages/react/menu/src/Menu.tsx#L544-L545) based on the decision we made when implementing the submenus. Do we think this should be different for non modal?

Comment (benoitgrelard) [reply packages/react/menu/src/Menu.tsx:300]: No I think it's all correct because menus should really be treated as one tab stop only.

**Comment by benoitgrelard** [comment]
@andy-hook this PR is now fully complete right? Just a note that I'd like to review and test this in it's entirety one last time.

File: packages/react/menu/src/Menu.tsx
Comment (jjenzz) [reply packages/react/menu/src/Menu.tsx:300]: Cool, if we're good with this it all looks fine to me :+1:

**Comment by jjenzz** [comment]
@benoitgrelard 

image

I assumed it was ready because of that ☝️

**Comment by andy-hook** [comment]
> this PR is now fully complete right? Just a note that I'd like to review and test this in it's entirety one last time before we merge.

Yessir. Sounds sensible.

File: .yarn/versions/ffb28263.yml
Comment (benoitgrelard) [inline .yarn/versions/ffb28263.yml:12]: Did we make sure this is up to date after all the PRs were merged into this one?

**Review by jjenzz — COMMENTED** [review]
You both know the inner workings on this better than me but just one thing on naming conventions:

File: .yarn/versions/ffb28263.yml
Comment (andy-hook) [reply .yarn/versions/ffb28263.yml:12]: Yes, it was updated just prior to the last PR, and that didn't introduce any version changes not already covered.

File: packages/react/context-menu/src/ContextMenu.tsx
Comment (benoitgrelard) [reply packages/react/context-menu/src/ContextMenu.tsx:199]: Haha, we literally changed them all the other day because I thought the opposite of what you're describing here. The biggest issue I have is that most of the time (in these at least) it's nearly impossible to name that thing with something actually accurate to "something that occured" because it encapsulates way too many cases. So you end up with a `hasInteractOutsideRef` which doesn't really mean that at all because we've grouped a bunch of different conditions inside it.

I rather in these cases prefer to do the commands thing because where the things happen we see what we're trying to do based on them.

Happy to re-consider if we find good names, but honestly there are some cases that will be impossible.

Comment (jjenzz) [reply packages/react/context-menu/src/ContextMenu.tsx:199]: > because we've grouped a bunch of different conditions inside it.

What are the conditions that are grouped inside it? As far as I can see they are if they interact outside and:

- it's not modal
- user hasn't prevented it

Both of these conditions when true mean that the user **hasn't** interacted outside because, they can't when it is modal or prevented.

But like I say, I don't care what we name it as long as it describes **what occurred** and not what it's going to control elsewhere because you lose context when reading the code. 

Everything needs code comments to describe the acceptance criteria rather than it just being self-documented.

Comment (benoitgrelard) [reply packages/react/context-menu/src/ContextMenu.tsx:199]: The worst offender that prompted it all for me was this one:

https://github.com/radix-ui/primitives/blob/4bb5024a2038df264a0bb6dfeb44819ff7776fc1/packages/react/dropdown-menu/src/DropdownMenu.tsx#L243-L253

Before this we ended up with something like:

```tsx
isInteractOutsideRef.current = !isModalLeftClick;
```

Something like that, and I just couldn't get my head around it, I find the new form much easier to understand.
But again, I'm not against doing everything back the other way if we can find a way that reads well but I haven't seen anything convincing yet.

Comment (andy-hook) [reply packages/react/context-menu/src/ContextMenu.tsx:199]: In that example I think it landed there because of the modal condition around the assignment, making the naming not technically accurate.

If we rejig it like this do you think it's clearer?

```jsx
onCloseAutoFocus={(event) => {
  props.onCloseAutoFocus?.(event);
  
  if (!event.defaultPrevented && !context.modal && hasInteractedOutsideRef.current) {
    event.preventDefault();
  }

  hasInteractedOutsideRef.current = false;
}}
onInteractOutside={(event) => {
  props.onInteractOutside?.(event);

  if (!event.defaultPrevented) hasInteractedOutsideRef.current = true;
}}
```

Comment (benoitgrelard) [reply packages/react/context-menu/src/ContextMenu.tsx:199]: Yeah that's exactly the point I made before your comment, you might have missed it @andy-hook.

Comment (andy-hook) [reply packages/react/context-menu/src/ContextMenu.tsx:199]: Oh snap I missed this entire thread in my stale browser window 🙈

Comment (jjenzz) [reply packages/react/context-menu/src/ContextMenu.tsx:199]: I think we're getting too caught up in the conditions. Just because I clicked outside doesn't mean I interacted outside. There are many variables that can prevent that "interaction" and they don't need to be part of the name, they are what determine its truthiness. 

An outside interaction only occurs if they haven't prevented it or if it's non-modal.

```jsx
const hasInteractedOutside = clickedOustide && !modal && !userPrevented;

// If whatever we consider an outside interaction **hasn't occurred* then prevent default
if (!event.defaultPrevented && !hasInteractedOutside) {
    event.preventDefault();
}
```

In the case of the right click stuff I would still just use `hasInteractedOutside` because that is part of what we consider an outside interaction in that case.

Comment (jjenzz) [reply packages/react/context-menu/src/ContextMenu.tsx:199]: To compare... 

```jsx
onClick={() => {
	if (!event.defaultPrevented) {
		hasClickedRef.current = true;
	}
}}
``

Are we implying this should actually be named `hasClickedAndUserDidntPreventItRef.current`? 

The latter part is irrelevant in my mind, it's what determines its value.

Comment (benoitgrelard) [reply packages/react/context-menu/src/ContextMenu.tsx:199]: I personally find it hard to envision anything without seeing an example, so I think at this point if we think we want to change things, I need to see the example re-written in the new way to see if I understand it.

Comment (jjenzz) [reply packages/react/context-menu/src/ContextMenu.tsx:199]: Sure, the original way @andy-hook wrote it in the PR is the way I am also describing which is quite telling that he did it that way off the bat too.

These changes got merged before I could comment on this but putting it back to something similar to how Andy originally did it gets my vote 👍

Comment (benoitgrelard) [reply packages/react/context-menu/src/ContextMenu.tsx:199]: I think writing it this way makes it a lot more prone to errors or missing a case, etc.
I like the explicitness of the cases where I can step through and understand why we do it.
I find it very confusing and couldn't get my head around it when I was reviewing it, but I'll let you guys settle things then.

**Review by jjenzz — APPROVED** [review]
I've given this a test locally, no obvious issues that I can find so we're gonna get this in Benoit for now. 

Definitely raise anything you find when you're back from holibobs though, we can fix separately.

**Review by andy-hook — APPROVED** [review]
Mirroring what @jjenzz said. It's funny to be in a situation where I'm approving my own changes again 😂

**Pull request merged**
